integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java

```diff
@@ -776,33 +793,127 @@ private int chownInternal(String path, long uid, long gid) {
+    return AlluxioFuseUtils.call(LOG, () -> truncateInternal(path, size),
+        "Fuse.Truncate", "path=%s,size=%d", path, size);
+  }
+
+  private int truncateInternal(String path, long size) {
+    // Truncate scenarios:
+    // 1. Truncate size = 0, file does not exist => no-op
+    // 2. Truncate size = 0, file exists and completed
+    // => noop if file size = 0 or delete file
+    // TODO(lu) delete open file entry if any
+    // 3. Truncate size = 0, file exists and is being written by current Fuse
+    // => noop if written size = 0, otherwise delete file and update create file entry
+    // 4. Truncate size = 0, file exists and is being written by other applications
+    // => error out
+    // 5. Truncate size != 0, file does not exist => error out
+    // 6. Truncate size != 0, file is completed
+    // => no-op if file size = truncate size, error out otherwise
+    // 7. Truncate size != 0, file is being written by other applications
+    // => error out, don't know exact file written size
+    // 8. Truncate size != 0, file is being written by current Fuse
+    // => no-op if written size = truncate size, error out otherwise
+    // Error out in all other cases
+    final AlluxioURI uri = mPathResolverCache.getUnchecked(path);
+    URIStatus status = null;
+    try {
+      status = mFileSystem.getStatus(uri);
+    } catch (FileDoesNotExistException | InvalidPathException e) {
+      status = null;
+    } catch (Throwable t) {
+      LOG.error("Failed to truncate path {} to {}. Failed to get file status", path, size, t);
+      return -ErrorCodes.EIO();
+    }
+    CreateFileEntry<FileOutStream> ce = mCreateFileEntries.getFirstByField(PATH_INDEX, path);
     if (size == 0) {
-      // truncate may be called in overwrite process:
-      // open(openflag=0b2) - truncate to size 0 - write - flush - release
-      if (!mCreateFileEntries.contains(PATH_INDEX, path)) {
-        LOG.error("Cannot truncate {} to {}. The file is not opened for overwrite", path, size);
+      if (status == null) {
+        // Case 1: Truncate size = 0, file does not exist => no-op
+        return 0;
+      }
+      try {
+        // Case 2: Truncate size = 0, file exists and completed
+        if (status.isCompleted()) {
+          if (status.getLength() == 0) {
+            return 0;
+          }
+          mFileSystem.delete(uri);
+          // TODO(lu) delete existing opened file entry if any
+          // require libfuse 3 which truncate() has fd info
+          // otherwise we need to change OpenFileEntries to include PATH_INDEX
+          // now fuse.read() will error out but it's acceptable
+          // because concurrent read and delete are not supported
+          return 0;
+        }
+        // Case 3: Truncate size = 0, file exists and is being written by current Fuse
+        if (ce != null) {
+          FileOutStream os = ce.getOut();
+          long bytesWritten = os.getBytesWritten();
+          if (bytesWritten == size) {
```

Comment:
```suggestion
          if (bytesWritten == 0) {
```

Since size == 0 as this line is within the `if (size == 0) `